src/main/java/com/truelayer/java/http/auth/AccessTokenInvalidator.java

```diff
@@ -16,10 +15,11 @@ public class AccessTokenInvalidator implements Authenticator {
     private final IAccessTokenManager tokenManager;
 
     @Override
-    public Request authenticate(Route route, Response response) throws IOException {
+    public Request authenticate(Route route, Response response) {
         if (isNotEmpty(response.request().header(AUTHORIZATION))) {
             // we are here in case of 401 on requests containing an Authorization HTTP header
-            tokenManager.invalidateToken();
+            String accessToken = response.request().header(AUTHORIZATION);
+            tokenManager.invalidateToken(accessToken.replace("Bearer", "").trim());
```

Comment:
This is not the more efficient solution as it will require iterating through the entire cache map entryset in order to find the value with the access token, but:
- it seems to me more explicit what we're trying to do
- the cache map won't contain lot of entries, i.e. as of now it will contain max 2 entries

Another approach could be to extract the `RequestScopes` tag from the request so that on the cache side we can hash it and directly obtain the cache key. In this way:
- it will be more efficient
- we remain consistent with the cache contract, as all the other methods take a `RequestScopes` as parameter

The main issue we may have with the second approach is that `hashcode()` method on `List` objects depends on the order, so there's the risk to fail to invalidate the token if the list of scopes is passed in a different order. This is an issue we may have as well when storing a new entry, maybe a possible solution could be to use a `Set` instead of a `List` in the `RequestScopes` object

I don't have a strong opinion tbh, I'm open for discussion

Reply:
Thanks for the investigation. I think I like the second option more, especially for your point on the contract consistency. Mixing RequestScopes and strings(access token) is a bit confusing IMO, and I think it would discourage from implementing a custom cache mechanism.

Moreover, I second your point on the number of entries and the opportunity to be pragmatic in this particular case, but the same conclusion that you had might not be applicable to other contexts as well. The current approach forces custom cache implementors to potentially list multiple entries on a shared cache (for instance a Redis server), and the assumption on a small number of entries can be not valid anymore. 

For these reasons I'd rather go with your last suggestion or KISS and invalidate the whole map if we can't do what you suggested. What you wrote makes a lot of sense to me, I'm just not sure on top of my head if we can access the scopes reliably [here](https://github.com/TrueLayer/truelayer-java/blob/ee6bbd706b9183871e87c2c18df5c179056da648/src/main/java/com/truelayer/java/http/auth/AccessTokenInvalidator.java#L22).

WRT your key instability doubt, I think we could sort the scopes naturally and then create the hash to have an idempotent mechanism for building the key. WDYT?